Update remote flow documentation, and add a picorv32/remote tutorial.

WRansohoff opened the pull request:
The new tutorial's verbiage could use a review, but it should be fairly complete.

aulihant commented on docs/tutorials/remote_soc.rst:
```diff
@@ -0,0 +1,258 @@
+You must have a recent version of Python3 and its Pip package manager in order to install SiliconCompiler. We recommend that Windows users download Python from `https://python.org/downloads/ <https://python.org/downloads/>`_, rather than the Microsoft Store.
+
+With that, you should be able to install SiliconCompiler directly from PyPi::
+
```
@WRansohoff and @gadfort : Is the expectation that all these tutorials will be completely standalone? That is - they should be able to skip our Installation instructions and walk through these instructions? If that is not the case, we should point back to our installation instructions to avoid duplicate or out-of-sync instruction.

WRansohoff replied:
Peter mentioned this in person today. I agree that the remote-specific instructions in the PicoRV32 tutorial should be moved to the remote section.

WRansohoff commented:
In response to feedback, I've moved most of the remote-specific explanations out of the PicoRV32 tutorial. I also removed some sections which won't be needed with the self-test target and default server address.

aulihant commented on docs/tutorials/remote_soc.rst:
Do we want the user to use our example (since the example without RAM is the same as what's here) or download from the YosysHQ GitHub repo that's pointed here?

WRansohoff replied:
Personally, I find it helpful to have a complete example to refer to when I follow tutorials that try to teach by building a project from the ground up.

When a bunch of small code blocks are interspersed with explanations, it can sometimes be hard to follow where exactly each one is supposed to go in the partially-completed example. So I thought it might be helpful to link to known-good examples for the designs that the tutorial covers.

aulihant replied:
Make sense. 
In that case, this will help make more clear this is the case, e.g, add:
> **This tutorial will walk you through downloading the design files from the source, but for your reference,** complete example designs which reflect the contents of this tutorial **are** in the public SiliconCompiler repository, [with RAM](https://github.com/siliconcompiler/siliconcompiler/tree/main/examples/picorv32_ram) and [without RAM](https://github.com/siliconcompiler/siliconcompiler/tree/main/examples/picorv32).

WRansohoff replied:
I ended up rewriting the paragraph a bit:

> We will walk through the process of downloading the design files and writing a build script, but for your reference, you can find complete example designs which reflect the contents of this tutorial in the public SiliconCompiler repository. The first part of the tutorial will cover building the CPU core [without RAM], and the second part will describe how to [add an SRAM block].

aulihant commented on docs/tutorials/remote_soc.rst:
I think you're missing:
`chip.input('picorv32.sdc')`

WRansohoff replied:
I'm planning to make a few changes which replace the SDC files with calls to `chip.clock()`

aulihant replied:
Do we want this to be consistent throughout all the examples? In many other places right now, it's using `chip.input('<design>.sdc'.

gadfort replied:
if the SDC is basically empty, then the .clock is probably better, otherwise we can use the SDC. The examples do not have to 100% consistent on this since we also need to show how to supply an SDC file

aulihant commented on docs/tutorials/remote_soc.rst:
```diff
@@ -0,0 +1,209 @@
+Building Your Own SoC
+=====================
+
+This tutorial will walk you through the process of building an ASIC containing one PicoRV32 RISC-V CPU core and 2 kilobytes of SRAM, on an open-source 130nm Skywater process node.
```
Explicitly add this is using remote flow:
> This tutorial will walk you through the process of building an ASIC containing one PicoRV32 RISC-V CPU core and 2 kilobytes of SRAM, on an open-source 130nm Skywater process node, **with SiliconCompiler's remote flow**.

aulihant commented on docs/tutorials/remote_soc.rst:
Add this to be explicit:
> and a report containing metrics related to the build, **in the `build/picorv32/<jobname>/` directory

aulihant commented on docs/tutorials/remote_soc.rst:
Add the file name since it is referenced directly later on:
> Then, create a Python script, **called ``sky130_sram_2k.py`` in that ``sram`` directory,** which describes...

aulihant commented on docs/tutorials/remote_soc.rst:
Later is referenced as `die_w`

WRansohoff replied:
Good catch, thanks - I think `width`/`height` are probably better for a tutorial since they're more descriptive than `w`/`h`.

aulihant commented on docs/tutorials/remote_soc.rst:
Later is referenced as `die_h`

aulihant commented on docs/tutorials/remote_soc.rst:
I think it would be helpful at this point, to provide what the directory tree looks like, for the user, as a sanity check to make sure all the files got copied over okay and named what the python script is expecting. 

This was my directory tree when I got it to work:
```
<rundir>
├── sram
│ ├── sky130_sram_2k.py
│ ├── sky130_sram_2kbyte_1rw1r_32x512_8.gds
│ └── sky130_sram_2kbyte_1rw1r_32x512_8.lef
├── picorv32.py
├── picorv32.sdc
├── picorv32.v
├── picorv32_top.py
├── picorv32_top.v
└── sky130_sram_2k.bb.v
```

WRansohoff replied:
Good idea. Maybe we should also have them put the SRAM blackbox file in the `sram/` directory?

aulihant replied:
It would be nicer, but I think that it's okay as is, for now, since this is working and matches with both the code and instructional text.
Maybe a nice-to-have for the next rev, and easier to track/follow encapsulated in its own PR.

WRansohoff replied:
Oh, whoops - I made the change in the last commit. It was quick because there aren't that many places where a `.bb.v` file extension is used.

aulihant commented on docs/tutorials/remote_soc.rst:
I think there is an indent missing here

WRansohoff replied:
You are correct. Sorry that I didn't actually follow this tutorial end-to-end before submitting it for review.

aulihant commented:
I followed the picorv32 tutorial all the way through the instructions and copied/pasted the files and ran. The good news is that with the minor notes I suggested for the copy/pasted files, it runs through!

gadfort commented on docs/index.rst:
```diff
@@ -37,6 +37,7 @@ Welcome to SiliconCompiler's documentation!
    :numbered:
 
    tutorials/zerosoc
+   tutorials/remote_soc
```
Lets move this below the frontend tutorials

gadfort commented on docs/tutorials/remote_soc.rst:
```diff
@@ -0,0 +1,220 @@
+
+See the :ref:`Installation <Installation>` section for information on how to install SiliconCompiler, and the :ref:`Remote Processing <Remote Processing>` section for instructions on setting up the remote workflow.
+
+[TODO / CR feedback: Add a screenshot here (or at the bottom)?]
```
I think you can add a screenshot here if you think it looks good and otherwise you can do it below as well

gadfort commented on docs/tutorials/remote_soc.rst:
If we have the file, I think we should reference the source like you have, but then provide a direct link to the file in our repo

gadfort commented on docs/tutorials/remote_soc.rst:
just refer to our file

gadfort commented on docs/tutorials/remote_soc.rst:
make a link to the skywater130 demo :ref:`skywater130-demo`

gadfort commented on docs/tutorials/remote_soc.rst:
We should make the "We are not returning the full results during this early beta period because we want to minimize bandwidth, and we believe that our public beta is currently best suited for rapid prototyping and design exploration." a .. note:: so people see it clearly (also add that to the remote processing).

gadfort commented on docs/tutorials/remote_soc.rst:
I think you can drop the "which have been published under a permissive license"

gadfort commented on docs/tutorials/remote_soc.rst:
Add direct link to sky130_sram_2kbyte_1rw1r_32x512_8, if possible

gadfort commented on docs/tutorials/remote_soc.rst:
Definitely add Screenshot here.

gadfort commented on docs/tutorials/remote_soc.rst:
Add Screenshot here.

gadfort commented on docs/user_guide/remote_processing.rst:
```diff
@@ -7,98 +7,51 @@ The SiliconCompiler project supports a remote processing model that leverages th
  #. NDA encumbered IPs, PDKs, and EDA tools.
 
-For command line compilation, remote processing is turned on with the '-remote' option. Results from a remote compilation should be identical to results from a local compilation.
+Note that our public beta currently only supports open-source tools and PDKs. You can access the public beta without a signup or login, and it is designed to delete your data after your jobs finish, but it is not intended to process proprietary or restricted intellectual property! Please `review our terms of service <https://www.siliconcompiler.com/terms-of-service>`_, and do not submit IP which  ...
```
Maybe put this in a .. note:: so it pops out

gadfort commented on docs/user_guide/remote_processing.rst:
```diff
@@ -7,98 +7,51 @@ The SiliconCompiler project supports a remote processing model that leverages th
+Note that our public beta currently only supports open-source tools and PDKs. You can access the public beta without a signup or login, and it is designed to delete your data after your jobs finish, but it is not intended to process proprietary or restricted intellectual property! Please `review our terms of service <https://www.siliconcompiler.com/terms-of-service>`_, and do not submit IP which  ...
+
+Even though our publicly-available servers only support open-source IP and tools, the remote API is capable of supporting any SiliconCompiler modules which the server operators wish to install, and we provide a minimal example development server which can be used as a starting point for custom server implementations. You can also find descriptions of the core remote API calls in the :ref:`remote  ...
+
```
Add comment about only returning the summary.

gadfort commented on docs/user_guide/remote_processing.rst:
```diff
@@ -7,98 +7,51 @@ The SiliconCompiler project supports a remote processing model that leverages th
+    sc -target asic_demo -remote
+
+The job should only take a few minutes to run if the servers aren't too busy, and once it completes, you should receive a PDF file containing a screenshot and metrics for the build results. The self-test design is a simple 8-bit counter, so your results should look like a small block of digital logic.
+
```
Add the screenshot here. That way you don't really need the "so your results should look like a small block of digital logic." part

gadfort commented on docs/user_guide/remote_processing.rst:
```diff
@@ -7,98 +7,51 @@ The SiliconCompiler project supports a remote processing model that leverages th
+Configuring a Different Remote Server
+-------------------------------------
+
+You can run the ``sc-configure`` command to configure your SiliconCompiler installation with a custom remote endpoint, if you have one. If your remote server requires authentication, you can run the utility with no arguments and fill in the address, username, and password fields that it prompts you for. If your server does not require authentication, you can simply pass its address in as a comman ...
```
:ref:`sc-configure`

gadfort commented on docs/user_guide/remote_processing.rst:
```diff
@@ -7,98 +7,51 @@ The SiliconCompiler project supports a remote processing model that leverages th
+Troubleshooting
+---------------
+
+Our public beta servers do not prune or pre-process Schema parameters, in order to make the remote processing environment as close to a local environment as possible. The jobs will be run in isolated environments with limited communication interfaces, however, so they jobs may not be able to perform network calls or meaningful filesystem access when they are run on our cloud beta servers.
```
"however, so they jobs may not be able to perform network calls or meaningful filesystem access when they are run on our cloud beta servers." this part is a little hard to follow, I think it's just missing some text

gadfort commented on docs/user_guide/remote_processing.rst:
```diff
@@ -7,98 +7,51 @@ The SiliconCompiler project supports a remote processing model that leverages th
+
+Our public beta servers do not prune or pre-process Schema parameters, in order to make the remote processing environment as close to a local environment as possible. The jobs will be run in isolated environments with limited communication interfaces, however, so they jobs may not be able to perform network calls or meaningful filesystem access when they are run on our cloud beta servers.
+
+Any changes that you make to SiliconCompiler's built-in tool setup scripts on your local machine will not be reflected in jobs which are run on a remote server. Likewise, any changes that you make to the built-in open-source PDKs and standard cell libraries will not be sent to the remote servers.
```
Maybe add a link to the contributions page and encourage them to contribute the changes

gadfort reviewed (changes requested):
Overall looks good, just some changes needed and maybe a few things we should chat about tomorrow

aulihant commented on docs/tutorials/remote_soc.rst:
What do you think about adding this? Might be helpful:
> Here is a shortcut to grab the lef and gds files if you find it helpful:
```
curl https://raw.githubusercontent.com/VLSIDA/sky130_sram_macros/main/sky130_sram_2kbyte_1rw1r_32x512_8/sky130_sram_2kbyte_1rw1r_32x512_8.gds > sky130_sram_2kbyte_1rw1r_32x512_8.gds
curl https://raw.githubusercontent.com/VLSIDA/sky130_sram_macros/main/sky130_sram_2kbyte_1rw1r_32x512_8/sky130_sram_2kbyte_1rw1r_32x512_8.lef > sky130_sram_2kbyte_1rw1r_32x512_8.lef

WRansohoff commented:
Okay, I think that I've applied all of our online and offline comments. In particular, I added screenshots where requested, marked the tutorial code blocks with captions, and did a quick visual review of the generated HTML pages.

gadfort commented:
I'm getting a: /tutorials/remote_soc.rst:24: WARNING: undefined label: 'skywater130-demo'

WRansohoff commented:
This PR was split into #1365 and #1361, which are both merged now.
